sync/sync.go

```diff
@@ -235,36 +239,129 @@ func (s *Synchronizer) verifierTask(ctx context.Context, block *core.Block, stat
 	}
 }
 
-func (s *Synchronizer) nextHeight() uint64 {
-	nextHeight := uint64(0)
-	if h, err := s.blockchain.Height(); err == nil {
-		nextHeight = h + 1
-	}
-	return nextHeight
-}
+
 
 func (s *Synchronizer) syncBlocks(syncCtx context.Context) {
 	defer func() {
 		s.startingBlockNumber = nil
 		s.highestBlockHeader.Store(nil)
 	}()
 
-	nextHeight := s.nextHeight()
+var nextHeight uint64
+	if h, err := s.blockchain.Height(); err == nil {
+		nextHeight = h + 1
+	} else {
+		nextHeight = 0
+	} 
 	startingHeight := nextHeight
 	s.startingBlockNumber = &startingHeight
-
+pendingSem := make(chan struct{}, 1)
 	latestSem := make(chan struct{}, 1)
 	if s.readOnlyBlockchain {
-		s.pollLatest(syncCtx, latestSem)
+		// Polling for the latest block in read-only mode
+		go func() {
+			poll := func() {
+				select {
+				case latestSem <- struct{}{}:
+					go func() {
+						defer func() { <-latestSem }()
+						highestBlock, err := s.starknetData.BlockLatest(syncCtx)
+						if err != nil {
+							s.log.Warnw("Failed fetching latest block", "err", err)
+						} else {
+							s.highestBlockHeader.Store(highestBlock.Header)
+						}
+					}()
+				default:
+				}
+			}
+
+			ticker := time.NewTicker(time.Minute)
+			defer ticker.Stop()
+			poll()
+
+			for {
+				select {
+				case <-syncCtx.Done():
+					return
+				case <-ticker.C:
+					poll()
+				}
+			}
+		}()
 		return
 	}
-
-	fetchers, verifiers := s.setupWorkers()
+numWorkers := 1
+	if s.catchUpMode {
+		m, mProcs := 16, runtime.GOMAXPROCS(0)
+		if mProcs > m {
+			numWorkers = m
+			} else {
+				numWorkers = mProcs
+			}
+		}
+	fetchers, verifiers := stream.New().WithMaxGoroutines(numWorkers), stream.New().WithMaxGoroutines(runtime.GOMAXPROCS(0))
 	streamCtx, streamCancel := context.WithCancel(syncCtx)
 
-	go s.pollLatest(syncCtx, latestSem)
-	pendingSem := make(chan struct{}, 1)
-	go s.pollPending(syncCtx, pendingSem)
+	go func() {
+		pollLatest := func() {
+			select {
+			case latestSem <- struct{}{}:
+				go func() {
+					defer func() { <-latestSem }()
+					highestBlock, err := s.starknetData.BlockLatest(syncCtx)
+					if err != nil {
+						s.log.Warnw("Failed fetching latest block", "err", err)
+					} else {
+						s.highestBlockHeader.Store(highestBlock.Header)
+					}
+				}()
+			default:
```

Comment:
Exercise 6 requires you to build a new, very simple, sync service. This requires you to retrieve data from the starknet sequencer, namely the new block, state update and any classes that have been declared that block. Once you have this data you should validate it. Once it's validated you can finally Store it in the database. And proceed to do the same for the next block etc.